client/comms/wsconn.go

```diff
@@ -254,17 +232,17 @@ func (conn *WsConn) read() {
 				log.Errorf("no handler found for response", string(b))
 				continue
 			}
-			handler.f(msg)
+			// Run handlers in a goroutine so that other messages can be recieved.
+			go handler.f(msg)
```

Comment:
The caller of `Request` needs to implement any necessary synchronization. Were you thinking of using a `WaitGroup` or something to track active handlers?

Reply:
Please check my understanding.
The client wants to know how much dcr they have in an order, and they want to save that data to a file. They would need to use a waitgroup inside the handler. If the caller just wants to print out that information, there is no need to wait, but whatever handler.f(msg) is doing will stop at some unknown point on context canceled.

Reply:
Good points here. I need to consider it a little more. I may butcher the terminology here, but **bbolt** is ACID and transactional, so I don't think we have any risk there. @chappjc could probably say more though.

Reply:
About `read()` in general, when the client shuts down, I believe something (in `wsConn`) should wait for `read()` to return, no? That's tricky since there doesn't seem to be a single run loop, rather `Connect()` provides a `WaitGroup` that only waits for `conn.close()` to return and while that triggers `read` to return it does not *wait* `for `read` to return.

As for the handlers, let me see if I understand how the pieces fit together: `core` or whoever will pass a function (the handler) to `comms.WsConn.Request` (a `comms.wsConn` method), which is registered in the `respHandlers` map and retrieved here in `read()`.

When underlying connection of the `wsConn` closes (either via the `close` method or otherwise), the `read` loop is going to return when `ReadJSON` returns an error.  If there are active handler goroutines running, that's fine, the handlers don't rely on the connection; they just process the response, right?

Reply:
Shows how much I know about databases. I though the idea of a "transaction" was that data integrity was ensured even in the event of catastrophic failure e.g. power outage.

Reply:
I'm just assuming the worst with bbolt.  It's possible that pulling the plug mid-write really is recoverable, but I wouldn't bank on it.  It's something to avoid if possible.

But this discussion (and the match DB work) sent me back to `Market` and `Swapper` to address some sloppy goroutine work on my part.  The new "epochPump" in https://github.com/decred/dcrdex/pull/145/commits/8b923cabd819d5f64ff2e3c5b24c317fc90d7cbb (PR #145) makes it fine to spend more time on `processEpoch` operations, including DB operations and the `Swapper.Negotiate` cal, which I previously fired off in unmonitored goroutines.  The changes, which were just leading up to the match DB work: https://github.com/decred/dcrdex/compare/master...chappjc:match-db-expanded?expand=1

Reply:
But what I really was getting at with those links to market and swap are that @JoeGruffins makes a good point that we need to consider if `main` could return back to the OS with running goroutines, and if any of those goroutines deserve to be supervised and allowed to gracefully return.  In most cases we should keep an eye on them.  If it's OK for a handler to abruptly terminate somewhere (anywhere) in its execution, then I think we should document that pretty clearly.  That applies to `read()` itself too.  Maybe it's fine, maybe not.